### ISSUE TITLE
Quoted string value with spaces does not work on command line but works in the shell

### ISSUE THREAD

**Issue by harendra-kumar** [hercules-team/augeas#301]
I am trying to change the eth0/pre-up to "sleep 2" but the value remains unchanged even though augtool's exit status is 0 and it does not print any error messages.

------snip-------
root@debian-jessie:~# augtool print /files/etc/network/interfaces/iface[2]/pre-up
/files/etc/network/interfaces/iface[2]/pre-up = "sleep 3"
root@debian-jessie:~# augtool set /files/etc/network/interfaces/iface[2]/pre-up "sleep 2"
root@debian-jessie:~# echo $?
0
root@debian-jessie:~# augtool print /files/etc/network/interfaces/iface[2]/pre-up
/files/etc/network/interfaces/iface[2]/pre-up = "sleep 3"
root@debian-jessie:~# 
------snip------

The same thing works through the augtool interactive shell though:

----snip-----
root@debian-jessie:~# augtool print /files/etc/network/interfaces/iface[2]/pre-up
/files/etc/network/interfaces/iface[2]/pre-up = "sleep 3"
root@debian-jessie:~# augtool << EOF                                             
> set /files/etc/network/interfaces/iface[2]/pre-up "sleep 2"
> save
> EOF
Saved 1 file(s)
root@debian-jessie:~# augtool print /files/etc/network/interfaces/iface[2]/pre-up
/files/etc/network/interfaces/iface[2]/pre-up = "sleep 2"
root@debian-jessie:~# 
----snip-----

augtool version is 1.2.0:

----snip----
root@debian-jessie:~# augtool --version
augtool 1.2.0 <http://augeas.net/>
Copyright (C) 2007-2011 David Lutterkort
License LGPLv2+: GNU LGPL version 2.1 or later
                 <http://www.gnu.org/licenses/lgpl-2.1.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by David Lutterkort
root@debian-jessie:~# 
----snip----

-harendra

**Comment by domcleal** [comment]
You need to call `save` for it to change the file.  Add the `-s` or `--autosave` argument to do this automatically, or use stdin or `--file` as you showed.

    augtool -s set /files/etc/network/interfaces/iface[2]/pre-up "sleep 2"

**Comment by harendra-kumar** [comment]
Thanks for a super quick response. I have two points in response to your
answer:

1) For a value without spaces in it I do not need "-s" or "--autosave". It
just saves the file without those options. I thought that's how it has been
designed. Is that a bug?

root@debian-jessie:~# augtool print
/files/etc/network/interfaces/iface[2]/pre-up

/files/etc/network/interfaces/iface[2]/pre-up = "sleep 2"

root@debian-jessie:~# augtool set
/files/etc/network/interfaces/iface[2]/pre-up 2

Saved 1 file(s)

root@debian-jessie:~# augtool print
/files/etc/network/interfaces/iface[2]/pre-up

/files/etc/network/interfaces/iface[2]/pre-up = "2"

2) My example does not work even with "-s" or "--autosave".  It displays
exactly the same behavior with or without these flags.

root@debian-jessie:~# augtool print
/files/etc/network/interfaces/iface[2]/pre-up

/files/etc/network/interfaces/iface[2]/pre-up = "2"

root@debian-jessie:~# augtool -s set
/files/etc/network/interfaces/iface[2]/pre-up "sleep 2"

root@debian-jessie:~# augtool print
/files/etc/network/interfaces/iface[2]/pre-up

/files/etc/network/interfaces/iface[2]/pre-up = "2"

root@debian-jessie:~# augtool --autosave set
/files/etc/network/interfaces/iface[2]/pre-up "sleep 2"

root@debian-jessie:~# augtool print
/files/etc/network/interfaces/iface[2]/pre-up

/files/etc/network/interfaces/iface[2]/pre-up = "2"

root@debian-jessie:~#


Another issue unrelated to the above is that it does not display any error
when the path specified is wrong. For example if I type "/file/..." instead
of "/files/...":

root@debian-jessie:~# augtool print
/file/etc/network/interfaces/iface[2]/pre-up

root@debian-jessie:~# echo $?

0

root@debian-jessie:~#

Here, I mistyped "files" (typed "file" instead) and it printed nothing. I
just kept wondering why its not working. Even set followed by save succeeds
on this incorrect path. Where does it save? I guess it just fails silently.
We can put some sanity checks to allow only recognised name spaces and
throw error on others to avoid the pain of discovering such problems the
hard way.

-harendra


PS: Thanks to the augeas team for creating such a wonderful tool.

**Comment by domcleal** [comment]
> 1) For a value without spaces in it I do not need "-s" or "--autosave". It just saves the file without those options. I thought that's how it has been designed. Is that a bug?
> root@debian-jessie:~# augtool
> set /files/etc/network/interfaces/iface[2]/pre-up 2
> Saved 1 file(s)

Yes, that looks like a bug.  https://github.com/hercules-team/augeas/blob/e034b42d0693fe7ce078c11396e90a1dd44697ea/src/augtool.c#L578 should be enclosed in braces for the if statement above it.



> 2) My example does not work even with "-s" or "--autosave". It displays exactly the same behavior with or without these flags.

It probably requires an extra level of quoting to work.  What's probably happening is that the set command is failing without it, for some reason augtool's not printing the resulting error message and exit code, but then nothing gets saved.

    augtool -s "set /files/etc/network/interfaces/iface[2]/pre-up 'sleep 6'"



> Another issue unrelated to the above is that it does not display any error when the path specified is wrong. For example if I type "/file/..." instead of "/files/...":

We've discussed adding a warning of some sort before, and I thought there was a ticket, but I can't find it right now.  There are legitimate reasons (e.g. the aug_store/retrieve APIs) to use parts of the tree outside of /augeas and /files though.

**Comment by harendra-kumar** [comment]
On Mon, Sep 28, 2015 at 3:40 PM, Dominic Cleal <notifications@github.com>
wrote:
>
> Yes, that looks like a bug.
> https://github.com/hercules-team/augeas/blob/e034b42d0693fe7ce078c11396e90a1dd44697ea/src/augtool.c#L578
> should be enclosed in braces for the if statement above it.
>
I wrote a few scripts on the assumption that just 'set' also saves when
used this way. I will fix those.  This should be fixed quickly otherwise
more people will be susceptible to script breakage when this bug gets fixed.

2) My example does not work even with "-s" or "--autosave". It displays
> exactly the same behavior with or without these flags.
>
> It probably requires an extra level of quoting to work. What's probably
> happening is that the set command is failing without it, for some reason
> augtool's not printing the resulting error message and exit code, but then
> nothing gets saved.
>
I verified it works if I provide the value as '"sleep 2"' i.e double quotes
inside single quotes. Which means, shell eats up the single quotes and
double quoted string gets passed to augtool. augtool eats up the double
quotes and the final value is set _without the quotes_ in it. But that is
strange and non-intuitive from a user perspective. It should not be
difficult to fix I guess.

> augtool -s "set /files/etc/network/interfaces/iface[2]/pre-up 'sleep 6'"
>
> Another issue unrelated to the above is that it does not display any error
> when the path specified is wrong. For example if I type "/file/..." instead
> of "/files/...":
>
> We've discussed adding a warning of some sort before, and I thought there
> was a ticket, but I can't find it right now. There are legitimate reasons
> (e.g. the aug_store/retrieve APIs) to use parts of the tree outside of
> /augeas and /files though.
>

Can we disallow unknown namespaces by default and allow them with a special
switch? Just an idea, I am not aware of the issues involved here. Currently
it is highly prone to silent errors if someone makes a spelling mistake.

-harendra

**Comment by domcleal** [comment]
> Yes, that looks like a bug.
>  https://github.com/hercules-team/augeas/blob/e034b42d0693fe7ce078c11396e90a1dd44697ea/src/augtool.c#L578
>  should be enclosed in braces for the if statement above it.
> I wrote a few scripts on the assumption that just 'set' also saves when
> used this way. I will fix those.  This should be fixed quickly otherwise
> more people will be susceptible to script breakage when this bug gets fixed.

Yes, I don't think we should change the behaviour at this point.  It's been that way since 1.0.0 (aa142d1a).
